### DISCUSSION TITLE
판매자 인증 토큰 발급 요청 오류

### DISCUSSION THREAD

**Discussion by kimdookyoung** [commerce-api-naver/commerce-solutions#55] [묻고 답하기]
안녕하세요

판매자가 커머스ID인증 이후 전달받은 토큰 JWE 해석 api 호출을 위해 인증 토근 발급 요청을 하는데요,
하기 처럼 작성하여 요청하였지만 client_secret_sign 항목이 유효하지 않습니다라고 나옵니다.

혹시 하기 내용중 암호화 방식이 잘못되었을까요??
$cryptedSignature = hash('sha256', $password . $clientSecret);

여러방법으로 계속 시도하고 있지만 같은 오류가 발생하고 있습니다...

도움 부탁드립니다~

$urlAccessToken = "https://api.commerce.naver.com/external/v1/oauth2/token";
$clientId = "애플리케이션 ID"; // 솔루션 애플리케이션 애플리케이션 ID
$clientSecret = '애플리케이션 시크릿'; // 솔루션 애플리케이션 애플리케이션 시크릿
$account_id = '판매자 ID'; //JWT 전달받은 accountUid 판매자 ID
$timestamp = (int)(microtime(true) * 1000);

/ 밑줄로 연결하여 password 생성
$password = $clientId . "_" . $timestamp;
$clientSecret = str_replace('+', '.', $clientSecret);

// 암호화
$cryptedSignature = hash('sha256', $password . $clientSecret);
$cryptedSignature = base64_encode($cryptedSignature);

$apiData = array(
    'client_id'          => $clientId,
    'timestamp'          => $timestamp,
    'client_secret_sign' => $cryptedSignature,
    'grant_type'         => 'client_credentials',
    'type'               => 'SELLER',
    'account_id'         => $account_id,
);

$curlHandle = curl_init();

$arr_headers = array("Content-Type: application/x-www-form-urlencoded");

$data = http_build_query($data);

curl_setopt($curlHandle, CURLOPT_POSTFIELDS, $data);
curl_setopt($curlHandle, CURLOPT_CUSTOMREQUEST, "POST");
curl_setopt($curlHandle, CURLOPT_URL, $urlAccessToken);
curl_setopt($curlHandle, CURLOPT_RETURNTRANSFER, true);
curl_setopt($curlHandle, CURLOPT_HTTPHEADER, $arr_headers);

/* CURL result */
$response = curl_exec($curlHandle);
curl_close($curlHandle);

response
{"code":"InvalidInputData","message":"입력한 데이터가 유효하지 않습니다.","timestamp":"2024-11-11T04:36:17.565822Z","invalidInputs":[{"name":"client_secret_sign","type":"NotValid","message":"client_secret_sign 항목이 유효하지 않습니다."}]}

**Comment by commerce-api-naver** [comment]
안녕하세요. 커머스API 이민석입니다.

인증 토큰 발급 과정에서 이용하는 전자서명(client_secret_sign)은 BCrypt 알고리즘을 이용합니다.
PHP의 경우 버전에 따라 crypt() 함수 혹은 password_hash() 함수를 널리 사용하는 것으로 알려져있습니다.
SHA-256이 아닌 BCrypt 알고리즘으로 시그니처를 생성하여 시도해보시기 바랍니다.
